src/mallocTracer.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "asprof.h"
+#include "assert.h"
+#include "codeCache.h"
+#include "mallocTracer.h"
+#include "os.h"
+#include "profiler.h"
+#include <dlfcn.h>
+#include <string.h>
+
+#define ADDRESS_OF(sym) ({               \
+    void* addr = dlsym(RTLD_NEXT, #sym); \
+    addr != NULL ? (sym##_t)addr : sym;  \
+})
+
+typedef void* (*malloc_t)(size_t);
+static malloc_t _orig_malloc = NULL;
+
+typedef void* (*calloc_t)(size_t, size_t);
+static calloc_t _orig_calloc = NULL;
+
+typedef void* (*realloc_t)(void*, size_t);
+static realloc_t _orig_realloc = NULL;
+
+typedef void (*free_t)(void*);
+static free_t _orig_free = NULL;
+
+static void* malloc_hook(size_t size) {
+    void* ret = _orig_calloc(1, size);
+    if (likely(ret && size)) {
+        MallocTracer::recordMalloc(ret, size, OS::nanotime());
```

Comment:
Why do you pass `OS::nanotime()` as argument every time?
`recordMalloc` can take a timestamp. A delay should be negligible.

Reply:
Delay wasn't negligible, `~50 ns` in my tests. This tiny delay caused a corresponding `free` to be recorded before `malloc` for some instances, breking the causality when events are sorted by time. I originally thought this was the cost of taking a stack trace.

Reply:
I mean, a timestamp can be taken in `MallocTracer::recordMalloc` before getting stack traces etc. There is nothing else in between - this should not take more than a couple of nanoseconds. In any case, I don't see how this can break the order, since `malloc` does not return before an event is emitted, and `free` is not called before `malloc` returns.

Reply:
free was not called before malloc, yet its timestamp was ~50ns lower than malloc's. Will take another look.